source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,134 @@ namespace aspect
+      {
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+            fe_values_12.reinit (cell);
+            fe_values_21.reinit (cell);
+
+            double local_area_integral = 0;
+
+            for (unsigned int q = 0; q < n_q_points_12; ++q)
+              {
+                local_area_integral += fe_values_12.JxW(q);
+              }
+            for (unsigned int c = 0; c < parameters.n_compositional_fields; ++c)
+              {
+                fe_values_12[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_12);
+                fe_values_21[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_21);
+
+                local_compositional_average_integrals[c] = 0;
+
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_compositional_average_integrals[c] += compositional_values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_compositional_average_integrals[c] /= local_area_integral;
+
+                double min_composition=compositional_values_12[0];
+                double max_composition=compositional_values_12[0];
+                for (unsigned int q = 1; q < n_q_points_12; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_12[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_12[q]);
+                  }
+                for (unsigned int q = 0; q < n_q_points_21; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_21[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_21[q]);
+                  }
+                // find the trouble cell
+                if (min_composition <composition_exact_min || max_composition >composition_exact_max)
+                  {
+                    //Define theta: which is scaling constant
+                    double theta_T=std::min<double>(1,abs((composition_exact_max-local_compositional_average_integrals[c])
+                                                          /(max_composition-local_compositional_average_integrals[c])));
+                    theta_T=std::min<double>(theta_T,abs((composition_exact_min-local_compositional_average_integrals[c])
+                                                         /(min_composition-local_compositional_average_integrals[c])));
+                    //Modify the numerical solution at freedoms
+                    double t_tmp=0;
```

Comment:
Please move the declaration of `t_tmp` inside the loop. It is only a double, and moving it inside the loop makes clear that every iteration is independent from the other iterations of the loop.